### PR TITLE
(WIP) Add tests to increase test coverage

### PR DESCRIPTION
**Pull request by albertferras-vrf**

### CODE REVIEW

**Comment by albertferras-vrf** [comment]
Hey @avilaton writting this here since I have no other way on discussing this with you :)
I'm taking a look at your PR https://github.com/dependabot/dependabot-core/pull/10040 and was checking on how to add tests to increase coverage to help you out.

I can see in `pip_compile_files` folder and `pip_compile_file_updater_spec.rb` there are many test situations with the normal `pip-compile` (non-uv) command, each tested and compared against the `requirements/something.txt`.
I believe the best thing would be to run all the same tests but using `uv`. However, duplicating each `something.in` into `something_uv.in` and another for requirements.txt in the same folder will make it a bit annoying to maintain both uv and normal pip-compile modes.

Did you want to test every single case or just a couple of them? If we want to test all cases maybe its better to duplicate `pip_compile_files`+`requirements` folders and create `uv_pip_compile_files` and `uv_requirements` folders, and also see if we can run `pip_compile_file_updater_spec.rb` to run with the 2 modes (trying to avoid duplicating this file..).

Also, not sure if the `.in` files should be duplicated. Can they be different for `uv` and `pip-compile` usage?

Thoughts?

**Comment by avilaton** [comment]
Hey, the only test I got around to writing payed off greatly and since this change wouldnaffect a large number of people, I want to be really careful.
I have no idea how much cover is good enough here, but clearly a single test isn't enough. 
Also, this is my first pr to dependabot, so I'm not quite sure.

**Comment by avilaton** [comment]
I think you are probably right about using the same .in files, not sure how to do it "the right way"® for the .txt files, would love to get some help from actual maintainers on it

**Comment by albertferras-vrf** [comment]
I have also never contributed to dependabot, and also don't have experience with Ruby :)
Apart from the header, do you know if using uv would give different results compared with pip-compile? Maybe for ~99% of the cases we can just reuse the same tests and expect both `.in` and `.txt` give the same results, but for very specific cases we can write one custom test for `uv`.
I don't know Ruby or its testing frameworks, but maybe it is easy to repeat the same tests with uv and pip-compile with just different .txt headers (the `autogenerated by uv`)

**Comment by avilaton** [comment]
Same with regards to ruby and this stuff, first time for me too.

Someone suggested we could use the next line from the header that reads `uv pip compile` instead which would be much more stable.

If this was python I could figure it out a bit better I guess.

**Comment by albertferras-vrf** [comment]
Seems like there is `Rspec.shared_examples` in ruby which lets you to re-run all tests twice but with distinct parameters.
I added two parameters, one with a distinct requirements folder (for uv) and one for the "autogen" header string that some tests check. Looks like this:

```
RSpec.describe Dependabot::Python::FileUpdater::PipCompileFileUpdater do
  context "with pip-compile" do
    it_behaves_like "PipCompileFileUpdater", "requirements/", "# This file is autogen"

  context "with uv" do
    it_behaves_like "PipCompileFileUpdater", "requirements_uv/", "# This file was autogenerated by uv"
  end
end
```

All tests are running twice now correctly, but there are ~10 which are failing (for uv). Some of them need to be disabled:
- .txt with hashes but without header (there is another test with header)
- '--legacy' resolver (afaik this is pip-compile specific)

So I tried to move those tests under the "with pip-compile" context thing, but errors start appearing because the `let` variables are not accessible in there 🤷 So not really sure how to do that.
If someone with very little ruby experience has to write this I guess it will be better to duplicate the tests+.txt files:
`uv_pip_compile_file_updater_spec.rb`
`requirements/uv_pip_compile_*.txt`

edit: I will do that, but creating a separate PR once I finish.

File: python/spec/dependabot/python/file_updater/pip_compile_file_updater_spec2.rb [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+
+      it "returns returns two hashes" do
+        result = updater.send(:package_hashes_for, name: name, version: version, algorithm: algorithm)
+        expect(result).to eq(["--hash=sha256:123abc"])
+      end
+    end
+  end
+end
+
+RSpec.describe Dependabot::Python::FileUpdater::PipCompileFileUpdater do
```
Comment (albertferras-vrf) [inline python/spec/dependabot/python/file_updater/pip_compile_file_updater_spec2.rb:630]: this is the fun part

**Pull request closed without merging**